### DISCUSSION TITLE
Bug in GCC when passing subarrays to procedures as `POINTER` dummy argument

### DISCUSSION THREAD

**Discussion by dev-zero** [cp2k/cp2k#2601] [Team Posts]
When working on some code I stumbled over a bug in GCC < 11:

  https://gcc.gnu.org/bugzilla/show_bug.cgi?id=85868 

In short: when passing a subarray of a `POINTER` to a procedure which takes the array as `POINTER`, GCC will generate wrong boundaries and indexes for that subarray.

The solution is to **NOT** declare array dummy arguments as `POINTER` unless you absolutely have to (like a re-allocation, checking whether it is associated or allocated).

Also, prefer `LBOUND/UBOUND` when iterating through an array and carefully check when changing array boundaries to an array (either explicitly in an allocation/declaration or implicitly when changing code to pass a subarray when it was the full array before).

As I had to figure out: neither valgrind nor `-fcheck` or `-fsanitize` or any compiler warnings help if you do read-only on that subarray.

Illustrated by the following snippet (see the invalid value and the bounds):

```
program main
    implicit none

    integer, dimension(:), pointer :: arr
    integer :: idx

    allocate(arr(3:13))
    arr = [1,2,3,4,5,6,7,8,9,0]

    call foo(arr)

    do idx = lbound(arr, 1), ubound(arr, 1)
        call foo(arr(idx:idx))
    end do

contains
    subroutine foo(a)
        integer, dimension(:), pointer, intent(in) :: a
        integer :: idx

        print *, "lbound/ubound:", lbound(a, 1), ubound(a, 1)
        do idx = lbound(a, 1), ubound(a, 1)
            print *, a(idx)
        end do
    end subroutine
end program
```

which yields (with GCC 10.2):

```
 lbound/ubound:           3          12
 val:           1
 val:           2
 val:           3
 val:           4
 val:           5
 val:           6
 val:           7
 val:           8
 val:           9
 val:           0
 lbound/ubound:           3           3
 val:           1
 lbound/ubound:           4           4
 val:           3
 lbound/ubound:           5           5
 val:           5
 lbound/ubound:           6           6
 val:           7
 lbound/ubound:           7           7
 val:           9
 lbound/ubound:           8           8
 val:      120529
 lbound/ubound:           9           9
 val:           0
 lbound/ubound:          10          10
 val:           0
 lbound/ubound:          11          11
 val:           0
 lbound/ubound:          12          12
 val:           0
 lbound/ubound:          13          13
 val:           0
```

Instead of the correct (GCC-trunk, Intel, Cray):

```
 lbound/ubound:           3          12
 val:           1
 val:           2
 val:           3
 val:           4
 val:           5
 val:           6
 val:           7
 val:           8
 val:           9
 val:           0
 lbound/ubound:           1           1
 val:           1
 lbound/ubound:           1           1
 val:           2
 lbound/ubound:           1           1
 val:           3
 lbound/ubound:           1           1
 val:           4
 lbound/ubound:           1           1
 val:           5
 lbound/ubound:           1           1
 val:           6
 lbound/ubound:           1           1
 val:           7
 lbound/ubound:           1           1
 val:           8
 lbound/ubound:           1           1
 val:           9
 lbound/ubound:           1           1
 val:           0
```

**Comment by oschuett** [comment]
Oh wow. This sounds like a really nasty pitfall. I guess, our strongest defense is testing with non-gnu compilers. 

I also agree that `POINTER` arguments are overused in CP2K. I'm wondering if we could somehow detect those cases?

**Comment by dev-zero** [comment]
> Oh wow. This sounds like a really nasty pitfall. I guess, our strongest defense is testing with non-gnu compilers.

Not necsssarily: depending on what is calculated based on that value, you may not see much difference. Example in my case was a neighbor list: For a small number of shells everything seemed to work (converging calculation, ok-ish value, etc.), it was only for a larger number of cells and/or more OpenMP threads that I started to occasionally got "broken" runs, which I then first attributed to some parallelization issue.

> I also agree that POINTER arguments are overused in CP2K. I'm wondering if we could somehow detect those cases?

Passing a subarray pointer should be visible in the AST, right? From there we should be able to get a list of procedures receiving a subarray, which we could then again check for their dummy argument declaration.

**Comment by mkrack** [comment]
Here is the behavior of various Intel compiler versions:

- Intel 18.0.5.274 (Build 20180823) and 19.0.3.199 (Build 20190206):
```
test.f90(13): error #7121: A ptr dummy may only be argument associated with a ptr, and this array element or section does not inherit the POINTER attr from its parent
        call foo(arr(idx:idx))
-----------------^
compilation aborted for test.f90 (code 1)
```
- Intel 19.0.4.243, 19.0.5.281, and 19.1.3.304:
```
 lbound/ubound:           3          13
           1
           2
           3
           4
           5
           6
           7
           8
           9
           0
           0
 lbound/ubound:           1           1
           1
 lbound/ubound:           1           1
           2
 lbound/ubound:           1           1
           3
 lbound/ubound:           1           1
           4
 lbound/ubound:           1           1
           5
 lbound/ubound:           1           1
           6
 lbound/ubound:           1           1
           7
 lbound/ubound:           1           1
           8
 lbound/ubound:           1           1
           9
 lbound/ubound:           1           1
           0
 lbound/ubound:           1           1
           0
```
It seems that the Intel compiler returns the right results starting from version 19.4 (same for 19.5 and 20.4).

**Comment by dev-zero** [comment]
@mkrack thanks for the tests, this means that if CP2K currently builds with Intel 18 then we're fine for now since the feature is not actively used yet, right? It's an F2008 feature, btw.

**Comment by mkrack** [comment]
@dev-zero yes, you are right, if we assume that the Intel compiler is reliable in detecting all these cases. The compiler version 18.0.5.274 is currently employed for the Intel regression test on Merlin listed in the dashboard.

**Comment by hfp** [comment]
If hoped for a long time to integrate CP2K on our side right our within compiler-lab. Let's see if this becomes true for Intel Next Gen (Fortran-)Compiler. I am close to believe it plays out...

**Comment by hfp** [comment]
Btw, was it working with IFORT 19.x?

**Comment by mkrack** [comment]
@hfp See above, 19.4, 19.5, and 20.4 work. 18.4 and 19.3 return a compilation error.

**Comment by hfp** [comment]
Thank you!
